### ISSUE TITLE
MCO Cluster Operator statuses frequently changing

### ISSUE THREAD

**Issue by kikisdeliveryservice** [openshift/machine-config-operator#346]
Unsure if this is causing any performance issues (it has not from what I can tell but)

The Cluster Operator machine-config-operator seems to cycle through statuses when the cluster is relatively idle (ie I am not performing and updates/applying mcs/etc...).  Not sure why the status is changing but it makes for a confusing user experience.  See the following.

`$ oc get clusteroperator -n machine-config-operator` yields varying statuses within very short periods of time.  For example:

`machine-config-operator                 3.11.0-494-g18b8dc50-dirty   True        False         False     7s`
`machine-config-operator                 3.11.0-494-g18b8dc50-dirty   False       True          False     1s`

In another example, I see changes in the `conditions:` section but the bottom entry also never changes? So the top two alternate between Available, False / Progressing, True but the oldest entry also never seems to change?

`$ oc get clusteroperator machine-config-operator -o yaml -w`
```
apiVersion: config.openshift.io/v1
kind: ClusterOperator
metadata:
  creationTimestamp: 2019-01-23T18:14:33Z
  generation: 1
  name: machine-config-operator
  resourceVersion: "919947"
  selfLink: /apis/config.openshift.io/v1/clusteroperators/machine-config-operator
  uid: bbc5a98d-1f3a-11e9-a4a4-02204cc1daaa
spec: {}
status:
  conditions:
  - lastTransitionTime: 2019-01-24T18:39:17Z
    status: "False"
    type: Available
  - lastTransitionTime: 2019-01-24T18:39:17Z
    message: Progressing towards 3.11.0-494-g18b8dc50-dirty
    status: "True"
    type: Progressing
  - lastTransitionTime: 2019-01-24T02:24:05Z
    status: "False"
    type: Failing
  extension:
    master: all 3 nodes are at latest configuration master-8b907698492f2829b35f6f826511dfa6
    worker: all 3 nodes are at latest configuration worker-58b30499568a7a2588b05ed01bba91a7
  version: 3.11.0-494-g18b8dc50-dirty
apiVersion: config.openshift.io/v1
kind: ClusterOperator
metadata:
  creationTimestamp: 2019-01-23T18:14:33Z
  generation: 1
  name: machine-config-operator
  resourceVersion: "919948"
  selfLink: /apis/config.openshift.io/v1/clusteroperators/machine-config-operator
  uid: bbc5a98d-1f3a-11e9-a4a4-02204cc1daaa
spec: {}
status:
  conditions:
  - lastTransitionTime: 2019-01-24T18:39:22Z
    status: "True"
    type: Available
  - lastTransitionTime: 2019-01-24T18:39:22Z
    status: "False"
    type: Progressing
  - lastTransitionTime: 2019-01-24T02:24:05Z
    status: "False"
    type: Failing
  extension:
    master: all 3 nodes are at latest configuration master-8b907698492f2829b35f6f826511dfa6
    worker: all 3 nodes are at latest configuration worker-58b30499568a7a2588b05ed01bba91a7
  version: 3.11.0-494-g18b8dc50-dirty
```
I'm also not seeing anything happening in the `oc logs -f` for the machine-config-operator at the times indicated in the statuses.

Does anyone have an idea why this is happening?  Why do the top two alternate, when from what I can tell nothing is happening in the MCO `oc logs -f`?

**Comment by kikisdeliveryservice** [comment]
cc @eparis 
I've seen the same behaviour that you describe and have opened this issue.

cc @abhinavdahiya i think this is from operator sync.go/status/go etc - do you have any idea what is happening?

**Comment by eparis** [comment]
While 'available' is false we won't be able to apply any cluster updates and the CVO won't be able to do it's job, right?

**Comment by kikisdeliveryservice** [comment]
Add: watching the `SINCE` column in `oc get clusteroperator -n machine-config-operator` I'm not seeing it make it over ~15s before it resets to 0s

**Comment by sjenning** [comment]
Below is a 2 minute watch at 1s intervals.

tl;dr the status is `Available` for ~20s, then moves to `Progressing`.  Status is updated again after about 5s but is still in `Progressing` state. Status is updated again after about 5s moving to `Available`.  It stays Available for about 20s again, then repeat.  The period of the cycle is ~30s, of which, 1/3 is spend in `Progressing` state.

```
$ oc get clusteroperators machine-config-operator; for i in $(seq 1 120); do oc get --no-headers=true clusteroperators machine-config-operator; sleep 1; done
NAME                      VERSION                      AVAILABLE   PROGRESSING   FAILING   SINCE
machine-config-operator   3.11.0-521-g28750463-dirty   True        False         False     8s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   8s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   10s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   11s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   12s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   14s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   15s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   16s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   17s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   19s <---
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   1s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   4s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   5s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   0s <---
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   3s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   4s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   1s <--
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   3s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   4s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   6s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   7s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   8s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   10s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   11s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   12s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   14s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   1s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   3s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   5s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   0s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   1s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   3s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   4s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   0s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   3s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   4s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   6s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   1s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   3s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   5s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   6s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   7s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   9s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   10s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   11s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   13s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   14s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   15s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   17s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   18s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   1s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   4s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   5s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   0s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   3s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   4s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   1s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   3s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   4s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   6s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   7s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   8s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   10s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   11s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   12s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   14s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   15s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   16s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   18s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   19s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   1s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   4s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   5s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   0s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   3s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   4s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   1s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   3s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   4s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   6s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   7s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   8s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   10s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   11s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   12s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   14s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   15s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   16s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   18s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   19s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   1s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   3s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   4s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   5s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   0s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   3s
machine-config-operator   3.11.0-521-g28750463-dirty   False   True   False   4s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   1s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   2s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   3s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   5s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   6s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   7s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   9s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   10s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   11s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   12s
machine-config-operator   3.11.0-521-g28750463-dirty   True   False   False   14s
```

**Comment by kikisdeliveryservice** [comment]
This issue should be fixed by #351 : The underlying cause was that the MCO was not adhering to CVO status rules. The "Available" status was erroneously tracking the "Progressing" status (which currently changes frequently due to MCO informer changes) instead of tracking whether the cluster is actually available at status: version X and functional.

Closing this issue now.  Thanks for your helpful logs and reports.  If you see any problems in the future, we will reopen.

**Comment by cgwalters** [comment]
Commenting on a closed issue, I know.  Not worth a new one yet.  I notice we still bounce into Progressing a lot seemingly for no reason.

**Comment by kikisdeliveryservice** [comment]
@cgwalters It will switch to Progressing whenever something is going on with any of the informers: https://github.com/openshift/machine-config-operator/pull/351#issuecomment-459070403

It happens so often though, I'm not sure it makes for a meaningful status.

**Comment by runcom** [comment]
This is supposed to fix the bouncing https://github.com/openshift/machine-config-operator/pull/406
